src/t8_forest/t8_forest_cxx.cxx

```diff
@@ -1372,55 +1555,127 @@ t8_forest_populate (t8_forest_t forest)
+  t8_element_t       *element;
+  t8_element_t       *tree_desc;
   t8_eclass_t         eclass;
   t8_eclass_scheme_c *ts;
+  t8_gloidx_t         global_neighbour_tree_idx;
   int                 ret;
-
-  T8_ASSERT (t8_forest_is_committed (forest));
-  T8_ASSERT (first_or_last == 0 || first_or_last == 1);
-  T8_ASSERT (forest != NULL);
-  if (forest->local_num_elements <= 0 || forest->trees == NULL
-      || forest->first_local_tree > forest->last_local_tree) {
-    /* This forest is empty and therefore the first tree is not shared */
+  int                 mpiret;
+  int                 mpirank_from;
+  int                 mpirank_to;
+  sc_MPI_Request      request;
+  sc_MPI_Status       status;
+
+  if (forest->mpisize == 1) {
+    /* Nothing to share */
     return 0;
   }
-  if (first_or_last == 0) {
-    /* Get a pointer to the first tree */
-    tree = (t8_tree_t) sc_array_index (forest->trees, 0);
-  }
-  else {
-    /* Get a pointer to the last tree */
-    tree = (t8_tree_t) sc_array_index (forest->trees,
-                                       forest->trees->elem_count - 1);
-  }
-  /* Get the eclass scheme of the first tree */
-  eclass = tree->eclass;
-  /* Get the eclass scheme of the first tree */
-  ts = t8_forest_get_eclass_scheme (forest, eclass);
-  /* Calculate the first/last possible descendant of the first/last tree */
-  /* we do this by first creating a level 0 child of the tree, then
-   * calculating its first/last descendant */
-  ts->t8_element_new (1, &element);
-  ts->t8_element_set_linear_id (element, 0, 0);
-  ts->t8_element_new (1, &desc);
-  if (first_or_last == 0) {
-    ts->t8_element_first_descendant (element, desc, forest->maxlevel);
+  if (forest->is_incomplete) {
+    if (first_or_last == 0) {
+      T8_ASSERT (forest->mpisize > 1);
+      if (forest->mpirank == 0) {
+        mpirank_from = forest->mpisize-1;
+        mpirank_to = forest->mpirank+1;
+      }
+      else if (forest->mpirank == forest->mpisize-1) {
+        T8_ASSERT (forest->mpirank > 0);
+        mpirank_from = forest->mpirank-1;
+        mpirank_to = 0;
+      }
+      else {
+        T8_ASSERT (forest->mpirank > 0 && forest->mpirank < forest->mpisize-1);
+        mpirank_from = forest->mpirank-1;
+        mpirank_to = forest->mpirank+1;
+      }
+      mpiret = sc_MPI_Irecv (&global_neighbour_tree_idx, 1, T8_MPI_GLOIDX,
+                             mpirank_from, 0, forest->mpicomm, &request);
+      SC_CHECK_MPI (mpiret);
+      mpiret = sc_MPI_Send (&forest->last_local_tree, 1, T8_MPI_GLOIDX,
+                            mpirank_to, 0, forest->mpicomm);
+      SC_CHECK_MPI (mpiret);
+      mpiret = sc_MPI_Wait (&request, &status);
+      SC_CHECK_MPI (mpiret);
+      if (!forest->mpirank) {
+        /* First process has nothing to do any more */
+        T8_ASSERT (!forest->mpirank);
+        return 0;
+      }
+      T8_ASSERT (global_neighbour_tree_idx < forest->global_num_trees); 
+      T8_ASSERT (global_neighbour_tree_idx > -1 ?
```

Comment:
Please document this non-trivial assertion and, if possible, split it into multiple ones.